pkg/config/v3alpha/config.go

```diff
@@ -0,0 +1,349 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v3alpha
+
+import (
+	"fmt"
+	"strings"
+
+	"sigs.k8s.io/yaml"
+
+	"sigs.k8s.io/kubebuilder/v3/pkg/config"
+	"sigs.k8s.io/kubebuilder/v3/pkg/model/resource"
+	"sigs.k8s.io/kubebuilder/v3/pkg/model/stage"
+)
+
+// Version is the config.Version for project configuration 3-alpha
+var Version = config.Version{Number: 3, Stage: stage.Alpha}
+
+type cfg struct {
+	// Version
+	Version config.Version `json:"version"`
+
+	// String fields
+	Domain     string `json:"domain,omitempty"`
+	Repository string `json:"repo,omitempty"`
+	Name       string `json:"projectName,omitempty"`
+	Layout     string `json:"layout,omitempty"`
+
+	// Boolean fields
+	MultiGroup      bool `json:"multigroup,omitempty"`
+	ComponentConfig bool `json:"componentConfig,omitempty"`
+
+	// Resources
+	Resources []resource.Resource `json:"resources,omitempty"`
+
+	// Plugins
+	Plugins PluginConfigs `json:"plugins,omitempty"`
+}
```

Comment:
@Adrio so far Kubebuilder exports the plugins field which allows SDK to do customizations on top. See: https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v3/memcached-operator/PROJECT#L12-L14

We need this in the interface otherwise it is a breaking change in the KB API for its consumer that does not allow the customizations be done.

Reply:
The `Plugin` field is exported because all the fields need to be exported in order to be marshalled to JSON/YAML. `Plugin` fields clearly specify how it should be used in its comments, stating that its goal is to store plugin configurations and that they should be encoded with the two provided methods. The methods documentation also provide more info on their usage. This exact same interface is implemented here.

SDK is using this field in a different way, storing a key of certain functionalities that were applid on the scaffolded files. These are not plugins as understood by kubebuilder nor are storing any configuration in this field. You can't say this is a breaking change when exactly the same interface as before is being provided just because a third-party was using it in a "hackish" way. The "hackish" way that is being used by SDK can still be used with th two provided methods, by trying to decode a certain plugin key. If the plugin key was not found it will yield an error while if found it will succeed.

Reply:
I could check. Thank you.